staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/connection.go

```diff
@@ -143,3 +156,24 @@ func (c *connection) newSpdyStream(stream *spdystream.Stream) {
 func (c *connection) SetIdleTimeout(timeout time.Duration) {
 	c.conn.SetIdleTimeout(timeout)
 }
+
+func (c *connection) sendPings(period time.Duration) {
+	t := time.NewTicker(period)
+	defer t.Stop()
+	for {
+		select {
+		case <-c.conn.CloseChan():
+			return
+		case <-t.C:
+		}
+		if _, err := c.conn.Ping(); err != nil {
+			klog.V(3).Infof("SPDY Ping failed: %v", err)
+			// Continue, in case this is a transient failure.
+			// c.conn.CloseChan above will tell us when the connection is
+			// actually closed.
+		} else {
+			// Count successful pings, for testing.
+			atomic.AddInt64(&c.pingsSent, 1)
```

Comment:
Can we avoid this testing-only code? Dose the SPDY library provide any testing support that exposes if a ping frame is received?

Reply:
Unfortunately, the library [handles incoming pings internally](https://github.com/docker/spdystream/blob/master/connection.go#L598-L607) without any external hooks I can find.
This unexported field is the least intrusive way I could verify the behavior. Using `sync/atomic`, performance overhead should be negligible.

Reply:
+1, if you want to test this, add a member variable of type function for calling Ping and let the test swap that out. It's not performance, it's just a code smell.

Reply:
fair enough, added a swap-able function field in `connection`